cmd/fanoutsidecar/main.go

```diff
@@ -125,57 +127,87 @@ func main() {
+func setupChannelWatcher(logger *zap.Logger, configUpdated swappable.UpdateConfig) (manager.Manager, error) {
 	mgr, err := manager.New(config.GetConfigOrDie(), manager.Options{})
 	if err != nil {
-		logger.Error("Error starting manager.", zap.Error(err))
+		logger.Error("Error creating new maanger.", zap.Error(err))
 		return nil, err
 	}
-
-	switch configMapNoticer {
-	case cmnfVolume:
-		err = setupConfigMapVolume(logger, mgr, configUpdated)
-	case cmnfWatcher:
-		err = setupConfigMapWatcher(logger, mgr, configUpdated)
-	default:
-		err = fmt.Errorf("need to provide the --config_map_noticer flag (valid values are %s)", configMapNoticerValues())
-	}
-	if err != nil {
+	if err = v1alpha1.AddToScheme(mgr.GetScheme()); err != nil {
+		logger.Error("Error while adding eventing scheme to manager.", zap.Error(err))
 		return nil, err
 	}
+	channelwatcher.New(mgr, logger, updateChannelConfig(configUpdated))
 
 	return mgr, nil
 }
 
-func setupConfigMapVolume(logger *zap.Logger, mgr manager.Manager, configUpdated swappable.UpdateConfig) error {
-	cmn, err := filesystem.NewConfigMapWatcher(logger, filesystem.ConfigDir, configUpdated)
-	if err != nil {
-		logger.Error("Unable to create filesystem.ConifgMapWatcher", zap.Error(err))
-		return err
-	}
-	if err = mgr.Add(cmn); err != nil {
-		logger.Error("Unable to add the config map watcher", zap.Error(err))
-		return err
+func updateChannelConfig(updateConfig swappable.UpdateConfig) channelwatcher.WatchHandlerFunc {
+	return func(ctx context.Context, c client.Client, chanNamespacedName types.NamespacedName) error {
+		channels, err := listAllChannels(ctx, c)
+		if err != nil {
+			logging.FromContext(ctx).Info("Unable to list channels", zap.Error(err))
+			return err
+		}
+		config := multiChannelFanoutConfig(channels)
+		return updateConfig(config)
 	}
-	return nil
 }
 
-func setupConfigMapWatcher(logger *zap.Logger, mgr manager.Manager, configUpdated swappable.UpdateConfig) error {
-	kc, err := kubernetes.NewForConfig(mgr.GetConfig())
-	if err != nil {
-		return err
+func listAllChannels(ctx context.Context, c client.Client) ([]v1alpha1.Channel, error) {
+	channels := make([]v1alpha1.Channel, 0)
+	for {
+		cl := &v1alpha1.ChannelList{}
+		opts := &client.ListOptions{
+			// Set Raw because if we need to get more than one page, then we will put the continue token
+			// into opts.Raw.Continue.
+			Raw: &metav1.ListOptions{},
+		}
+		if err := c.List(ctx, opts, cl); err != nil {
+			return nil, err
+		}
+		for _, c := range cl.Items {
+			if c.Status.IsReady() && shouldWatch(&c) {
+				channels = append(channels, c)
+			}
+		}
+		if cl.Continue != "" {
+			opts.Raw.Continue = cl.Continue
+		} else {
+			return channels, nil
+		}
 	}
+}
 
-	cmw, err := watcher.NewWatcher(logger, kc, configMapNamespace, configMapName, configUpdated)
-	if err != nil {
-		return err
+func shouldWatch(ch *v1alpha1.Channel) bool {
+	if ch.Spec.Provisioner != nil && ch.Spec.Provisioner.Namespace == "" {
```

Comment:
Maybe make namespace non-empty an early exit with a comment that we only support cluster-scoped provisioners (so it's easier to extend to namespace-scoped provisioners later)?

Reply:
Didn't understand the comment. This function is used to filter out channels in the handler inside the watch. It is not used in the request path.

Reply:
I was suggesting writing this as:

```golang
if ch.Spec.Provisioner == nil || ch.Spec.Provisioner.Namespace != "" {
  // Only support cluster-level provisioners right now.
  return false
}
...